### PR TITLE
Checking for exercise errors

### LINKED ISSUE #356: Enable tutorial.error.checker to provide checks on error
If a student submits code that throws an error, learnr exits the checking routine and displays the error message. This short circuit prevents the checking code from telling the student anything useful about what might have gone wrong. In short, the student is left to decipher the error message on his or her own.

This is necessary when the checking code relies on the result of the student code (if it throws an error, there is no result). But it is a missed opportunity when the checking code parses the unevaluated student code, as `gradethis::grade_code()` does. Here is an example, `grade_code()` would return `"I expected mtcars where you wrote mt"` but learnr displays the error message. (code appended at bottom)

Screen Shot 2020-04-23 at 2 58 41 PM

To fix this, I suggest that we:
1. Add a `tutorials.error.checker` option that can be set to a grading function to run in the event that the student code throws an error. This option can be set in the same manner as the `tutorials.exercise.checker` option. It should be set to a function that can grade the student code without evaluating it, similar to `gradethis::grade_code()`.
Screen Shot 2020-04-23 at 3 19 10 PM

 
2. Display the result of the error checking function beneath the error message. Here is a mock up. 
error-mockup 001

Code to recreate existing example above
****
````
---
title: "Example"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
library(gradethis)

tutorial_options(exercise.checker = gradethis::grade_learnr)
knitr::opts_chunk$set(out.width = "100%")
```

## Welcome

### `mtcars`

To see the `mtcars` data frame, type `mtcars` in the code block below and click "Submit Answer".

```{r ex, exercise = TRUE}

```

```{r ex-solution}
mtcars
```

```{r ex-check}
grade_code("Good job!")
```
````

### PR DESCRIPTION
**Pull request by cpsievert**
Closes #356. Note that this allows checking of (evaluation) errors without any additional options. Here's a minimal example:

````
---
title: "Checking errors"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)	
library(gradethis)
# remotes::install_github("rstudio-education/gradethis")
tutorial_options(
  exercise.checker = grade_learnr, 
  exercise.error.checker = grade_learnr_error
)
```

* Throw an error message of `"boom"`:

```{r grade_error, exercise = TRUE}

```

```{r grade_error-solution}
stop("boom")
```

```{r grade_error-code-check}
grade_code("Nice error!")
```
````

### Known breaking change

Now, if a `-code-check` chunk exists, then we no longer display the results of the exercise (press "Run Code" if you want results and/or do the grading in the `-check` chunk)

### TODO

- [x] Update news (and mention breaking change)

### CODE REVIEW

File: R/options.R [hunk 1]
```diff
@@ -22,9 +24,12 @@ tutorial_options <- function(exercise.cap = "Code",
                              exercise.timelimit = 30,
                              exercise.lines = NULL,
                              exercise.checker = NULL,
+                             exercise.error.checker = NULL,
+                             exercise.parse.checker = NULL,
```
Comment (cpsievert) [inline R/options.R:28]: This option really wants to live in gradethis, but it doesn't seem worth re-exporting

Comment (cpsievert) [reply R/options.R:28]: Remove the changes here in `R/options.R`

File: R/exercise.R [hunk 2]
```diff
@@ -176,215 +176,144 @@ evaluate_exercise <- function(exercise, envir, evaluate_global_setup = FALSE) {
...
-  err <- NULL
-  get_checker <- function() {
-    checker <- exercise$options$exercise.checker
-    if (is.function(checker)) {
-      environment(checker) <- envir_prep
-    } else if (!is.null(checker)) {
-      warning("Found a exercise.checker that isn't a function", call. = FALSE)
-      checker <- NULL
+  # Setup a temporary directory for rendering the exercise
+  exercise_dir <- tempfile(pattern = "learnr-tutorial-exercise")
```
Comment (cpsievert) [inline R/exercise.R:191]: Because Windows

```suggestion
  exercise_dir <- tempfile(pattern = "lnr-ex")
```

File: R/options.R [hunk 3]
```diff
@@ -22,6 +22,7 @@ tutorial_options <- function(exercise.cap = "Code",
                              exercise.timelimit = 30,
                              exercise.lines = NULL,
                              exercise.checker = NULL,
+                             exercise.error.checker = NULL,
```
Comment (schloerke) [inline R/options.R:25]: Needs param line above

**Pull request merged**